darshan-runtime/lib/darshan-core.c

```diff
@@ -759,39 +756,1085 @@ void darshan_core_shutdown(int write_log)
         }
 #endif
 
-        darshan_core_fprintf(stderr, "#darshan:<op>\t<nprocs>\t<time>\n");
-        darshan_core_fprintf(stderr, "darshan:log_open\t%d\t%f\n", nprocs, open_tm);
-        darshan_core_fprintf(stderr, "darshan:job_write\t%d\t%f\n", nprocs, job_tm);
-        darshan_core_fprintf(stderr, "darshan:hash_write\t%d\t%f\n", nprocs, rec_tm);
-        darshan_core_fprintf(stderr, "darshan:header_write\t%d\t%f\n", nprocs, header_tm);
-        for(i = 0; i < DARSHAN_MAX_MODS; i++)
+        darshan_core_fprintf(stderr, "#darshan:<op>\t<nprocs>\t<time>\n");
+        darshan_core_fprintf(stderr, "darshan:log_open\t%d\t%f\n", nprocs, open_tm);
+        darshan_core_fprintf(stderr, "darshan:job_write\t%d\t%f\n", nprocs, job_tm);
+        darshan_core_fprintf(stderr, "darshan:hash_write\t%d\t%f\n", nprocs, rec_tm);
+        darshan_core_fprintf(stderr, "darshan:header_write\t%d\t%f\n", nprocs, header_tm);
+        for(i = 0; i < DARSHAN_MAX_MODS; i++)
```

Comment:
DARSHAN_MAX_MODS is not safe for this loop, because the i variable is being used as an index into darshan_module_names[], which is declared with an x macro and only has DARSHAN_KNOWN_MODULE_COUNT entries in it.  Might need to audit and see if there are any other loops like this where this could be a problem.

Or maybe a safer solution would be a macro that looks up the name of a module based on it's id and returns a null string or "UNKNOWN" if the id > KNOWN_MODULE_COUNT.  That way we wouldn't have to worry about manually indexing for the x macro arrays that have a smaller size.

Reply:
I can open an issue to look more at this. There do appear to be 1 or more spots in the runtime code that could be problematic, but I don't see any introduced in this PR. Actually, in the `dump_config()` routine, I'm surprised I remembered to do the right thing with `DARSHAN_KNOWN_MODULE_COUNT`.

Reply:
Opened #712 to track this